### ISSUE TITLE
feature(cli): Extend and implement ng2-seed-cli

### ISSUE THREAD

**Issue by d3viant0ne** [mgechev/angular2-seed#609]
This is something I have been thinking about after playing around with https://github.com/AngularClass/NG6-starter

The thought resurfaced after a discussion with @Meligy in #605 and I think it would be a beneficial addition to the seed.

**Proposal**
- Add a template based component generator
- Allow for configuration of the css pre/post processor
- Allow for a custom template set seed / project tooling standard
- Wiki documentation for custom templates

It wouldn't be a great deal of work and I feel there would be a definite benefit in velocity / development experience.

@mgechev @ludohenin @nareshbhatia @NathanWalker opinions?

After discussion - We will be extending https://github.com/ludohenin/ng2-seed-cli to cover the customizations we support with the seed.

**Comment by TheDonDope** [comment]
Hi @d3viant0ne ,

if i may, i want to add my perspective on this topic

I had some previous experience with Ruby on Rails, and what absolutely loved the experience of scaffolding a new resource, by simple doing something like `$ rails scaffold User name:string password:string` and getting all the basic components as well as models and controllers for the different CRUD operations and basic views. 
Then after playing around a bit with Angular2, scaffolding was the one thing i found lacking and would love to have something like this. 
I would agree it would make a great fit for this seed project here, and implementation as a seed specific helper task would not be that complicated.

Have a nice day

Greetings

Christian

**Comment by mgechev** [comment]
@d3viant0ne seems good to me. It'll be even better if it is not part of the seed officially. May be an optional npm package, which could be installed and adds some extra gulp tasks for component/service/etc/ scaffolding?

**Comment by d3viant0ne** [comment]
Yeah, I'm down for an angular2-seed plugin :)

**Comment by ludohenin** [comment]
@d3viant0ne I've a prototype for that, not sure what it worth but you should maybe have a look.
https://github.com/ludohenin/ng2-seed-cli

**Comment by d3viant0ne** [comment]
@ludohenin - That's certainly a solid jumping off point. It really just needs the ability to add custom templates and a few other "nice to have" items.

I'll tinker around with it this weekend.

**Comment by ludohenin** [comment]
@mgechev Would you create a repo `angular2-seed-cli` ?

**Comment by d3viant0ne** [comment]
^^ The would certainly keep things nice and tidy.

**Comment by NathanWalker** [comment]
@d3viant0ne I really like this idea. My only suggestion would be to make the component generator customizable so if the project used decorators for instance, the generator could use a component template that would already be setup to use that decorator. This would be useful for many other things as well. Additionally a generator for the unit test (and that would allow for a custom template as well - some test setups may need certain boilerplate across different projects and using a custom test template to generate from would help).
If not too difficult; that would be really excellent.

**Comment by d3viant0ne** [comment]
@NathanWalker The plan is to allow for user defined custom templates or a set of defaults based on the style-guide.

**Comment by mgechev** [comment]
@d3viant0ne @ludohenin can we use [this repository](https://github.com/ludohenin/ng2-seed-cli) for the purpose?

**Comment by ludohenin** [comment]
@mgechev @d3viant0ne Totally fine with me :)

**Comment by d3viant0ne** [comment]
@mgechev - I think it would be better to keep everything seed related in one place. Less searching, easier to find related items.

Perhaps you should create an org for things like this? Optionally, I have one that everyone can have owner access to ( it just exists to make managing access easier )

**Comment by mgechev** [comment]
@d3viant0ne sounds good to me, I'm just not sure how much traction an eventual cli will get given that `angular-cli` already exists. My suggestion is to keep working into a separate repository and once we find out that additional tools around the seed get traction, to move it to an organization.

**Comment by d3viant0ne** [comment]
@mgechev - I agree. I think the plan is to add update tools to the cli + some basic dev. experience tools around the ability to add components with the latter more like a stand in until the cli is ready.

**Comment by ludohenin** [comment]
so we go for ng2-seed-cli ?

**Comment by d3viant0ne** [comment]
@ludohenin - Yeah, let's go to work.

**Comment by ludohenin** [comment]
@mgechev @d3viant0ne I've granted you access to [this](https://github.com/ludohenin/ng2-seed-cli) repo.

**Comment by Meligy** [comment]
Hello,

The idea of handling some scaffolding for things like controllers / routes is awesome.

With the introduction of CLI tool, it might be useful to put light to a couple of questions:

- Wouldn't a CLI tool be directly duplicating the angular-cli effort This angular-seed project solves the same problem, but differently. With CLI, it will be solving the same problem, pretty much the same way (on the high-level).
  
  - Would it make sense / be possible, to use the angular-cli for this somehow?

    That would be possible if the controller scaffolders are not tied to the broccoli / ember-cli stuff in angular-cli of course, otherwise likely not possible.

  - If not, would it make sense to make this entire project cli-centric?

    If it's a duplicate anyway.

- Would it make sense to utilize an existing templating infrastructure for the scaffold? Namely create it as yeoman generator?

    Is this already in the cards?

    It kinds reflects on whether the project would benefit from being cli/generator centric, although both questions are not necessarily correlated.

Thanks a lot.

**Comment by NathanWalker** [comment]
I have found that angular-cli project is not planning on becoming *extensible*. From what I've encountered, sounds like their scaffolding will always be specific to using their build setup so may not be the best choice when providing scaffolding here.

I agree with the yeoman generator sentiment. May be possible to join efforts. This looked promising months ago: 
https://www.npmjs.com/package/generator-an2
https://medium.com/@houssemyahiaoui/angularjs-2-yeoman-generator-an2-ff7f705b3f80#.9t3wh556c
Maybe contact author to continue developing if worth expanding to support more generators?
